### PR TITLE
Amy

### PR DESCRIPTION
**Pull request by amymariewall**

### CODE REVIEW

**Comment by amymariewall** [comment]
Link to Phase 1 Google Workbook so far. Am I on the right track? https://docs.google.com/spreadsheets/d/1IWrRWfLzcV8QT0s7N3JjA4mYjCIQMlteDiyvwTc2wew/edit?usp=sharing

**Comment by sumeetjain** [comment]
- Events - It seems like you're storing the same information in multiple columns. Do you need to have "100m" in the event's name, if that's also represented by the event's distance?
- Is the 'styles' table something you know for sure you'll need, based on the project specification?
- Same question as above for lane assignments.
- Your 'wins' table should have a column to indicate the "rank" of a particular win (since prizes for 1st through 3rd are given for each event)

**Comment by amymariewall** [comment]
Thank you for the feedback! 
- Events - good point, I changed the names to be completely separate from any other information.
- Styles - No, it is not needed explicitly for the assignment, it was just something I included while thinking over the logistics of what the competition administrator _might_ need. Should I remove it for simplicity?
- Lane assignments - Same thought process there. I think the Lane Assignment table can be removed and instead I could use the 'finishes' table only.
- I added a rank to the wins table, thank you for that!

**Comment by sumeetjain** [comment]
:+1: 

> Styles - No, it is not needed explicitly for the assignment, it was just something I included while thinking over the logistics of what the competition administrator might need. Should I remove it for simplicity?

Nope, it's a cool feature. Just wanted to check if I'd miscommunicated something.

**Comment by amymariewall** [comment]
Meagan and I partnered up and we switched it up a bit to be a speech and debate tournament. We followed the same structure of before for the most part. Does this still look good? https://docs.google.com/spreadsheets/d/1IWrRWfLzcV8QT0s7N3JjA4mYjCIQMlteDiyvwTc2wew/edit?usp=sharing

**Comment by sumeetjain** [comment]
- Please describe "max_possible_score_per_student" – not sure I follow.
- Please describe "meetups".
- Yes, it could be possible to simply query the results table to get wins instead of having a wins table. We'll see how things look to me after you've clarified the above though.

**Comment by amymariewall** [comment]
- max_possible_score_per_student is the highest possible score a student could get for that activity type. So, for example, in an Original Oratory Event, the highest possible score a student could get would be 6, but in a Model UN Event, the highest possible score a student could get would be 5.
- meetups are "events" - the debate equivalent of individual 'heats' in a swim meet.

**Comment by maedae** [comment]
To add on that, basically our "winners" could be set on a singular event level, but it might be better to do a query based on the total accumulated score for a student in a certain activity (they could be doing two student congress sessions), divided by the total amount of possible points across those multiple activities -- or even total accumulated points across all participated events, divided by the total possible points one could acquire in all those events. 

Generally, though -- the more events you do, the better off you are because you have the chance to acquire more points. There is also the possibility of actually being able to score negative points if you did poorly.

**Comment by sumeetjain** [comment]
Let's favor simplicity a bit more, please. I suggest:
- [ ] Remove the 'events' table. It's an extra layer of complexity between students and the activities they compete in. Just have students competing in activities–nevermind when that particular activity occurs. Just pool everyone into one big competitive group.
- [ ] Keep the algorithm for winners as simple as you can manage.

**Comment by amymariewall** [comment]
Okay - I think we were thinking of having multiple events of each activity type in the tournament. So the "activities" in our table would be like "style" for swimming - breaststroke, freestyle, backstroke, etc. Then there are multiple 'events' of the same style....

Is that too complicated?

**Comment by amymariewall** [comment]
We just decided to simplify - so we will remove 'events' and just have 'activities'. There will only one of each activity in the competition. Thanks for the suggestion!

**Comment by maedae** [comment]
In that same vein, would it be better to also just put the student score and ID on the activity table and drop the results table too?

**Comment by maedae** [comment]
updated doc:
https://docs.google.com/spreadsheets/d/1IWrRWfLzcV8QT0s7N3JjA4mYjCIQMlteDiyvwTc2wew/edit?usp=sharing

**Comment by sumeetjain** [comment]
> would it be better to also just put the student score and ID on the activity table and drop the results table too?

No. You'll still want a clean way to say which students are competing in which events. Certainly more than one student will compete in an event, so if you drop the results table then you'll get a lot of repetition in the activities table.

**Comment by maedae** [comment]
Okay, cleaned it based on what we just talked about. Anything else you would recommend tweaking or looking into adding ahead of time?

**Comment by sumeetjain** [comment]
:+1:

File: app/views/admin/add_result.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<h1>Add Result:</h1>
+
+<h2>Please enter the result:</h2><br>
+<form action="/result/create">
+
+Student:<br>
+
+<select name="student_id">
+ <% @students.each do |s| %>
```
Comment (sumeetjain) [inline app/views/admin/add_result.erb]: Indentation is off.

File: app/views/admin/add_result.erb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  <option value="<%= s.id %>"><%= s.first_name %><%= s.last_name %></option>
+  <% end %>
+</select><br>
+
+Activity:<br>
+<select name="activity_id">
+ <% @activities.each do |a| %>
```
Comment (sumeetjain) [inline app/views/admin/add_result.erb]: Indentation is off.

File: app/views/admin/add_student.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+Student Last Name:<br>
+<input type="text" name="student_last_name"><br>
+Student Age:<br>
+<input type="number" name="student_age"><br>
+
+Student College:<br>
+
+<select name="college_id">
+ <% @colleges.each do |c| %>
+  <option value="<%= c.id %>"> <%= c.college_name %></option>
```
Comment (sumeetjain) [inline app/views/admin/add_student.erb]: Indentation is off.

File: partner.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Name: Meagan Jungman
+GitHub Username: maedae
```
Comment (sumeetjain) [inline partner.md]: Partner file should be deleted.

**Comment by amymariewall** [comment]
Thank you! I modified my indentations and reorganized the controller to more closely follow the assignment instructions. Submitting again for approval up to this point.

**Comment by sumeetjain** [comment]
:+1:

**Comment by amymariewall** [comment]
I would like to TIAS on some ways to make tables communicate with each other, but want to get approval on what I've done so far first. 

I think ActiveRecord allows you to tell it how the tables are related to each other using methods, and then you can call a table on a model object. So in my Student model I could define a method 
`has_many : results`, 
and then I could define methods where I call `results` on a Student object:
`student_result_in_event = student.results.find_by_event_id(event_object)`
and easily organize ranks and scores for students and events...but I need to explore it a little more to figure out exactly how it works.

**Comment by sumeetjain** [comment]
> has_many : results

That _does_ work. But it's a shortcut that I'd like you to _not_ use until we go over it. It's important that you first learn to write methods for accessing other tables your self.

**Comment by sumeetjain** [comment]
Otherwise, this code it looking :+1:

**Comment by amymariewall** [comment]
Sounds good, thanks! I will resist trying to use that for now.

**Comment by amymariewall** [comment]
I am submitting for approval through Phase 5 - If I don't stop now, I'm pretty sure I will just keep adding new features forever...

**Comment by amymariewall** [comment]
I modified my registration page to have a checkbox where you can add a student to multiple events at once. I am currently using dropdowns to select which student the admin wants to register or edit. 

I really enjoy working with these UX components, and I have implemented them a lot in my app. I would really like to investigate more about how to inform one dropdown's options by the selection the user makes in another dropdown...but I won't do so if you prefer that I stay away from that for now.

**Comment by sumeetjain** [comment]
Nice work!

> I won't do so if you prefer that I stay away from that for now.

Yes please.

**Comment by amymariewall** [comment]
Drat.

Fair enough, I shall work on improvements. Thanks!

**Pull request closed without merging**